pkg/provisioner/demo/demo.go

```diff
@@ -182,6 +185,27 @@ func (p *demoProvisioner) UpdateHardwareState() (result provisioner.Result, err
 	return result, nil
 }
 
+// Adopt allows an externally-provisioned server to be adopted.
+func (p *demoProvisioner) Adopt() (result provisioner.Result, err error) {
+	hostName := p.host.ObjectMeta.Name
+
+	switch hostName {
+	default:
+		p.host.ObjectMeta.Name = AdoptingHost
```

Comment:
Why change the object name here?

Reply:
To demonstrate the multi-stage nature of the adoption (i.e. it's not complete after the first call to Adopt()).

Reply:
The point of the demo driver is to ensure that we can create a bunch of example host CRs that will get "stuck" in certain states, so that someone (esp. the UI team) could use the API to see what they look like in those cases. So, rather than changing the host name, I think we want to just return the same stable state over and over for the hosts based on their name (even if that causes reconciliation again).

The fixture driver is the place to deal with the fact that we expect adoption to require multiple iterations, so that the unit tests will take the host through the expected flow.

Reply:
It would also be good to update deploy/crds/demo-hosts.yaml with the new cases.

Reply:
Got rid of this in the new patches. Do we still need to do anything in demo-hosts.yaml?

Reply:
Yeah, ideally we would add 1 new host names to the list of special names. "adopting-host" would always appear to be in the process of being adopted and would stay dirty (so the demo driver would always return dirty when the name matches). We can do that in a separate PR, I think.

Reply:
Oh, gotcha. I'm not sure that's interesting from a UI perspective though, because the visible state of the Host doesn't actually change - it just sits in ExternallyProvisioned either way and this is more of an implementation detail, sorting out the bookkeeping in Ironic.
Anyway, I agree we can deal with that in a separate PR if we need to.